src/gtk/toga_gtk/widgets/detailedlist.py

```diff
@@ -1,92 +1,231 @@
+
+    def row_factory(self, item):
+        """
+        Args:
+            item (:obj:`Row`)
+        Returns:
+            Returns a (:obj:`TextIconRow`)
+        """
+        new_item = TextIconRow(self.interface.factory, self, item)
+        item._impl = new_item
+        return new_item
+
+    def destroy(self):
+        self.disconnect(self.gtk_on_select_signal_handler)
+        super().destroy()
 
-        self.treeview.set_model(self.store)
+    def change_source(self, source):
+        """
+        Args:
+            source (:obj:`ListSource`)
+        """
+        self.store.remove_all()
+        for item in source:
+            self.store.append(
+                self.row_factory(item))
+
+        # We can't know the dimensions of each row (and thus of the list) until gtk allocates
+        # space for it. Gtk does emit `size-allocate` after allocation but I couldn't find any
+        # guarantees that the rows have their sizes allocated in the order they are inserted
+        # in the `ListStore` and in my opinion that's unlikely to be the case.
+
+        # Therefore we would need to wait for `size-allocate` on all rows and either update
+        # the visibility of the buttons on all `size-allocates` or have a counter and only do
+        # it on the last `size-allocate`. Obviously none of those options are desirable.
+
+        # Fortunately functions added with `idle_add` are run when gtk is idle and thus after
+        # any size allocation. This solves our problem and from the perspective of the user
+        # happens immediately.
+
+        # Even though we are adding the callback to the global loop, it only runs once.
+        # This is what the lambda is for. If a callback returns `False` then it's not ran again.
+        # I used a lambda because returning `False` from `self._list_items_changed()` would mean
+        # returning `False` on success.
+        GLib.idle_add(lambda: not self._list_items_changed())
 
     def insert(self, index, item):
-        self.store.insert(item)
+        """
+        Args:
+            index (int)
+            item (:obj:`Row`)
+        """
+        new_item = self.row_factory(item)
+
+        if index == -1:
+            self.store.append(new_item)
+        else:
+            self.store.insert(index, new_item)
+
+        self.list_box.show_all()
+        self._list_items_changed()
 
     def change(self, item):
-        self.store.change(item)
+        """
+        Args:
+            item (:obj:`Row`)
+        """
+        new_item = self.row_factory(item)
+        index = item._impl.get_index()
+
+        if index == -1:
+            self.store.append(new_item)
+        else:
+            self.store.insert(index, new_item)
 
     def remove(self, item, index):
-        self.store.remove(item, index)
+        """
+        Removes a row from the store. This method doesn't trigger the `on_delete` callback and it
+        doesn't remove the row from the interface.
```

Comment:
err... it doesn't? Why not? What removes the row from the interface?

Reply:
The user, by calling `dl.data.remove()`, which in turn calls `dl_gtk.remove()`. The remove function should call `on_delete()` though.

This brought another problem to my attention: When a row is removed gtk immediately calls `dl_gtk.gtk_on_row_selected()`, and thus the `dl.on_select()` callback.
On the detailedlist example, at the bottom of the window, it briefly displays "No row selected" and then "Row English is going to be deleted.". This is probably always going to work because signals are very fast, but signals are also asynchronous so there is no guarantee.